packages/accounts-password/password_server.js

```diff
@@ -782,71 +864,22 @@ Meteor.methods({resetPassword: function (token, newPassword) {
+ * @param {Object} [extraTokenData] Optional additional data to be added into the token record.
+ * @returns {Object} Object with {email, user, tokenRecord, url, options} values.
  * @importFromPackage accounts-base
  */
-Accounts.sendVerificationEmail = function (userId, address) {
+Accounts.sendVerificationEmail = function (userId, email, extraTokenData) {
   // XXX Also generate a link using which someone can delete this
   // account if they own said address but weren't those who created
   // this account.
 
-  // Make sure the user exists, and address is one of their addresses.
-  var user = Meteor.users.findOne(userId);
-  if (!user)
-    throw new Error("Can't find user");
-  // pick the first unverified address if we weren't passed an address.
-  if (!address) {
-    var email = _.find(user.emails || [],
-                       function (e) { return !e.verified; });
-    address = (email || {}).address;
-
-    if (!address) {
-      throw new Error("That user has no unverified email addresses.");
-    }
-  }
-  // make sure we have a valid address
-  if (!address || !_.contains(_.pluck(user.emails || [], 'address'), address))
-    throw new Error("No such email address for user.");
-
-
-  var tokenRecord = {
-    token: Random.secret(),
-    address: address,
-    when: new Date()};
-  Meteor.users.update(
-    {_id: userId},
-    {$push: {'services.email.verificationTokens': tokenRecord}});
-
-  // before passing to template, update user object with new token
-  Meteor._ensure(user, 'services', 'email');
-  if (!user.services.email.verificationTokens) {
-    user.services.email.verificationTokens = [];
-  }
-  user.services.email.verificationTokens.push(tokenRecord);
-
-  var verifyEmailUrl = Accounts.urls.verifyEmail(tokenRecord.token);
-
-  var options = {
-    to: address,
-    from: Accounts.emailTemplates.verifyEmail.from
-      ? Accounts.emailTemplates.verifyEmail.from(user)
-      : Accounts.emailTemplates.from,
-    subject: Accounts.emailTemplates.verifyEmail.subject(user)
-  };
-
-  if (typeof Accounts.emailTemplates.verifyEmail.text === 'function') {
-    options.text =
-      Accounts.emailTemplates.verifyEmail.text(user, verifyEmailUrl);
-  }
-
-  if (typeof Accounts.emailTemplates.verifyEmail.html === 'function')
-    options.html =
-      Accounts.emailTemplates.verifyEmail.html(user, verifyEmailUrl);
-
-  if (typeof Accounts.emailTemplates.headers === 'object') {
-    options.headers = Accounts.emailTemplates.headers;
-  }
-
+  const {email: realEmail, user, tokenRecord} =
+    Accounts.generateVerificationTokenRecord(userId, email, extraTokenData);
+  const userWithTokenRecord = Accounts.saveVerificationTokenRecord(user, tokenRecord);
+  const url = Accounts.urls.verifyEmail(tokenRecord.token);
+  const options = Accounts.generateOptionsForEmail(realEmail, userWithTokenRecord, url, 'verifyEmail');
   Email.send(options);
+  return {email: realEmail, user: userWithTokenRecord, tokenRecord, url, options};
```

Comment:
This seems to be more complicated than the original design I thought we were working off of [from here](https://github.com/meteor/meteor/pull/8065#issuecomment-263877938).

Why do we need to expose the new `saveVerificationTokenRecord` as a public API (and most importantly the additional maintenance cost)?  Can't it just be done inside `Accounts.generateVerificationTokenRecord`?